R/case-variables.R

```diff
@@ -0,0 +1,154 @@
+#' (useful when made into a case variable)
+#' missing a logical indicating if this case should be treated as missing
+#' 
+#' @keywords internal
+ensureValidCases <- function(case) {
+    if (!is.list(case)) {
+        halt("A case must be a list")
+    }
+    wrong_case_names <- setdiff(names(case), c("id", "name", "expression",
+                                               "numeric_value", "missing"))
+    if (length(wrong_case_names) > 0) {
+        halt("each case must have at most an id, name, expression, ",
+             "numeric_value, and missing element. The errant arguments were: ",
+             serialPaste(wrong_case_names))
+    }
+    if (!is.null(case$id) && !is.integer(case$id)) {
+        halt("a case's id must be an integer")
+    }
+    if (!is.character(case$name)) {
+        halt("a case's name must be a character")
+    }
+    if (class(case$expression) != "CrunchLogicalExpr") {
+        halt("a case's expression must be a CrunchLogicalExpr")
+    }
+    if (is.null(case$numeric_value)) {
+        case['numeric_value'] <- list(NULL)
+    } else if (!is.numeric(case$numeric_value)) {
+        halt("a case's numeric_value must be a numeric")
+    }
+    if (is.null(case$missing)) {
+        case['missing'] <- FALSE
+    } else if (!is.logical(case$missing)) {
+        halt("a case's missing must be a logical")
+    }
+    
+    return(case)
+}
+
+#' Make a case variable
+#' 
+#' The `makeCaseVariable` function derives a variable using values from other 
+#' variables. These are evaluated in the order they are supplied in the list 
+#' as the `cases` argument (they proceed in an IF, ELSE IF, ELSE IF, ..., ELSE 
+#' fashion); the first one that matches selects the corresponding value from 
+#' the case list.
+#' 
+#' There are two ways to specify cases, but you must pick only one (note these 
+#' two will produce the same case variable):
+#' 
+#' 1. When you just want to specify conditions, you can use named conditions:
+#' `makeCaseVaraible(case1=ds$v1 == 1, case2=ds$v2 == 2, name="new case")`
+#' 
+#' 1. You can also use the `cases` argument, which is useful when you want to 
+#' prespecify ids, or set numeric_values:
+#' `makeCaseVaraible(cases=list(list(expression=ds$v1 == 1, name="case1"), list(expression=ds$v2 == 2, name="case2")), name="new case")`
+#' 
+#' @param ... a sequence of named expressions to use as cases as well as other 
+#' properties to pass about the case variable (i.e. alias, description)
+#' @param cases a list of lists with each case condition to use each must 
+#' include at least a `name` and an `expression` element.
+#' @param else_case a single list that has at least a `name` to serve as the 
+#' case when no others match (if not specified Crunch will use the system 
+#' default "No data")
+#' @param name a character to use as the name of the case variable to create
+#' @export
+makeCaseVariable <- function (..., cases, else_case, name) {
+    dots <- list(...)
```

Comment:
You could probably do `casevar <- list(..., name=name)` and save yourself two lines later (and get your input validation about name being required done in the process).